pyproject.toml

```diff
@@ -1,3 +1,73 @@
+[project]
+name = "ops"
+description = "The Python library behind great charms"
+readme = "README.md"
+requires-python = ">=3.8"
+authors = [
+    { name="The Charm Tech team at Canonical Ltd.", email="charm-tech@lists.launchpad.com" },
+]
+classifiers = [
+    "Programming Language :: Python :: 3",
+    "License :: OSI Approved :: Apache Software License",
+    "Development Status :: 5 - Production/Stable",
+    "Intended Audience :: Developers",
+    "Intended Audience :: System Administrators",
+    "Operating System :: MacOS :: MacOS X",
+    "Operating System :: POSIX :: Linux",
+]
+dependencies = [
+    "PyYAML==6.*",
+    "websocket-client==1.*",
+]
+dynamic = ["version"]
+
+[project.optional-dependencies]
+docs = [
+    "sphinx==6.2.1",
+    "sphinx-design",
+    "furo",
+    "sphinx-tabs",
+    "lxd-sphinx-extensions",
+    "sphinx-copybutton",
+]
+dev = [
+    "isort~=5.11",
+    "autopep8~=1.6",
+    "flake8~=6.1",
+    "flake8-docstrings~=1.7",
+    "flake8-builtins~=2.1",
+    "pyproject-flake8~=6.1",
+    "pep8-naming~=0.13",
+    "pytest~=7.2",
+    "pyright==1.1.317",
+    "pytest-operator~=0.23",
+    "coverage[toml]~=7.0",
+    "typing_extensions~=4.2",
+]
+
+[project.urls]
+"Homepage" = "https://juju.is/docs/sdk"
+"Repository" = "https://github.com/canonical/operator"
+"Issues" = "https://github.com/canonical/operator/issues"
+"Documentation" = "https://ops.readthedocs.io"
+"Changelog" = "https://github.com/canonical/operator/blob/main/CHANGES.md"
+
+[build-system]
+requires = [
+    "setuptools>=60",
+    "PyYAML==6.*",
+    "websocket-client==1.*",
```

Comment:
Let me check this.

Reply:
I don't think there is no. Normally, you wouldn't need to have the project dependencies as build dependencies. We have this issue because we're using a dynamic `version`, and that version is found by the build system looking for `ops.__version__`, and if you import `ops` then you import (basically) _everything_ because (almost) everything's in the top level name space, so you need to have everything installed. It could look for `ops.version.version` but that still imports the package, so has the same issue.

I think the 'modern' approach is meant to be that instead of doing something like `ops.__version__` you do `importlib.metadata.version('ops')`, although you need to have installed it for that to work, and we can't do require people to this because it'd break backwards compatibility.

I tried to find other projects that have a dynamic version field and similar issues to see if someone else has an elegant solution, but wasn't successful. I couldn't figure out anything myself either. So I think we are stuck with it, annoyingly.

Reply:
Per discussion, we can avoid the project deps if we use `ops.version.version`, because the dynamic metadata tool [is a bit clever](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata) and tries to load it using the AST and `ast.literal_eval` first, then uses import as a fallback.

Reply:
Ben found a trick that I missed :tada: - the setuptools docs say:

> It is advisable to use literal values together with attr (e.g. str, tuple[str], see [ast.literal_eval()](https://docs.python.org/3.11/library/ast.html#ast.literal_eval)). This is recommend in order to support the common case of a literal value assigned to a variable in a module containing (directly or indirectly) third-party imports.
>
> attr first tries to read the value from the module by examining the module’s AST. If that fails, attr falls back to importing the module

So `ops.__version__` falls back to importing because it's imported from another module. But `ops.version.version` is a literal value, so that works without needing to import (I just verified this). So will switch to that, thanks!